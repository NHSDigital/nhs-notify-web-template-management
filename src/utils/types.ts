export type FormId =
  | 'choose-a-template-type'
  | 'create-nhs-app-template'
  | 'create-email-template'
  | 'create-letter-template'
  | 'create-text-message-template'
  | 'preview-nhs-app-template'
  | 'submit-template'
  | 'create-nhs-app-template-back';

export type FormErrorState = {
  formErrors: string[];
  fieldErrors: Record<string, string[]>;
};

export enum TemplateType {
  NHS_APP = 'NHS_APP',
  SMS = 'SMS',
  EMAIL = 'EMAIL',
  LETTER = 'LETTER',
}

export type Session = {
  id: string;
  templateType: TemplateType | 'UNKNOWN';
  nhsAppTemplateName: string;
  nhsAppTemplateMessage: string;
};

export type FormState = {
  validationError?: FormErrorState;
  redirect?: string;
};

export type TemplateFormState = FormState & Session;

export type PageProps = {
  params: {
    sessionId: string;
  };
};

export type TemplateSubmittedPageProps = {
  params: {
    templateId: string;
  };
};

export type PageComponentProps = {
  initialState: TemplateFormState;
};

export type SubmitTemplatePageComponentProps = {
  templateName: string;
  sessionId: string;
};

export enum TemplateFormatText {
  APP = 'APP',
  EMAIL = 'EMAIL',
  LETTER = 'LETTER',
  SMS = 'SMS',
}
