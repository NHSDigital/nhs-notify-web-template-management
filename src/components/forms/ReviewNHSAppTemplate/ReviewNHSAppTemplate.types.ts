export type ReviewNHSAppTemplateProps = {
  templateName: string;
  message: string;
};
