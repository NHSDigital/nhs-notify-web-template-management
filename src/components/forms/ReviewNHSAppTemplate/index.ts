export * from './ReviewNHSAppTemplate';
export * from './server-action';
