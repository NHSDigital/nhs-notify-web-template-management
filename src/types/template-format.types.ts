export enum TemplateFormatText {
  APP = 'APP',
  EMAIL = 'EMAIL',
  LETTER = 'LETTER',
  SMS = 'SMS',
}
