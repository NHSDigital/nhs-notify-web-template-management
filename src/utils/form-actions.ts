'use server';

import { getAmplifyBackendClient } from '@utils/amplify-utils';
import { DbOperationError } from '@domain/errors';
import { Template, Draft } from './types';
import { logger } from './logger';

export async function createTemplate(
  template: Draft<Template>
): Promise<Template> {
  const { data, errors } =
    await getAmplifyBackendClient().models.TemplateStorage.create(template);

  if (errors || !data) {
    logger.error('Failed to create template', errors);
    throw new Error('Failed to create new template');
  }
  return data;
}

export async function saveTemplate(template: Template): Promise<Template> {
  const { data, errors } =
    await getAmplifyBackendClient().models.TemplateStorage.update(template);

  if (errors) {
    logger.error('Failed to save template', errors);
    throw new Error('Failed to save template data');
  }

  if (!data) {
    throw new DbOperationError({
      message:
        'Template in unknown state. No errors reported but entity returned as falsy',
      operation: 'update',
    });
  }
  return data;
}

export async function getTemplate(
  templateId: string
): Promise<Template | undefined> {
  const { data, errors } =
    await getAmplifyBackendClient().models.TemplateStorage.get({
      id: templateId,
    });

  if (errors) {
    logger.error('Failed to get template', errors);
  }

  if (!data) {
    logger.warn(`Failed to retrieve template for ID ${templateId}`);
    return undefined;
  }

  return data;
}

export async function sendEmail(
  templateId: string,
  templateName: string,
  templateMessage: string,
  templateSubjectLine: string | null
) {
  const res = await getAmplifyBackendClient().queries.sendEmail({
    recipientEmail: 'england.test.cm@nhs.net',
    templateId,
    templateName,
    templateMessage,
    templateSubjectLine,
  });

  if (res.errors) {
    logger.error({
      description: 'Error sending email',
      res,
    });
  }
}

export async function getTemplates(): Promise<Template[] | []> {
  const { data, errors } =
    await getAmplifyBackendClient().models.TemplateStorage.list();

  if (errors) {
    logger.error('Failed to get templates', errors);
  }

  if (!data) {
    logger.warn(`Failed to retrieve templates`);
    return [];
  }

  return data;
}
