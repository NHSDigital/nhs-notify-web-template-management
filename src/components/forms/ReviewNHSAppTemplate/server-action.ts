import { z } from 'zod';
import { MarkdownItWrapper } from '@utils/markdownit';
import { TemplateFormState } from '@utils/types';
import { SendEmail } from '@utils/form-actions';

export function renderMarkdown(
  value: string,
  markdown = new MarkdownItWrapper()
) {
  markdown.enableLineBreak().enable(['heading', 'link', 'list', 'emphasis']);

  return markdown.render(value);
}

const radioSelectionToPageMap: Record<'nhsapp-edit' | 'nhsapp-submit', string> =
  {
    'nhsapp-edit': 'create-nhs-app-template',
    'nhsapp-submit': 'submit-template',
  };

const schema = z.object({
  reviewNHSAppTemplateAction: z.enum(['nhsapp-edit', 'nhsapp-submit'], {
    message: 'Select an option',
  }),
});

export async function reviewNhsAppTemplateAction(
  formState: TemplateFormState,
  formData: FormData
): Promise<TemplateFormState> {
  const form = Object.fromEntries(formData.entries());
  const validationResponse = schema.safeParse(form);

  if (!validationResponse.success) {
    return {
      ...formState,
      validationError: validationResponse.error.flatten(),
    };
  }

  await SendEmail();

  const page =
    radioSelectionToPageMap[validationResponse.data.reviewNHSAppTemplateAction];

  return {
    ...formState,
    redirect: `/${page}/${formState.id}`,
  };
}
