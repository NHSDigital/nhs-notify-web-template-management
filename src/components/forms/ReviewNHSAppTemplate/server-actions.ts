import { MarkdownItWrapper } from '@/src/utils/markdownit';
import { FormState } from '@utils/types';

export function renderMarkdown(
  value: string,
  markdown = new MarkdownItWrapper()
) {
  markdown.enableLineBreak().enable(['heading', 'link', 'list', 'emphasis']);

  return markdown.render(value);
}

export function handleForm(state: FormState, _data: FormData) {
  return { ...state };
}
