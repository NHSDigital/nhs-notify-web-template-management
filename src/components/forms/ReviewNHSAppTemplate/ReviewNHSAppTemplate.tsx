'use client';

import { PreviewTemplate } from '@molecules/PreviewTemplate';
import { ReviewTemplate } from '@organisms/ReviewTemplate';
import { PageComponentProps } from '@utils/types';
import content from '@/src/content/content';
import { renderMarkdown } from './server-actions';
import { NHSNotifyBackButton } from '@molecules/NHSNotifyBackButton/NHSNotifyBackButton';

export function ReviewNHSAppTemplate({ state, action }: PageComponentProps) {
  const { nhsAppTemplateName, nhsAppTemplateMessage } = state;

  const html = renderMarkdown(nhsAppTemplateMessage);

  const {
    components: { previewNHSAppFormComponent },
  } = content;

  return (
    <div className='nhsuk-grid-row'>
      <NHSNotifyBackButton
        formId='review-nhs-app-template-back'
        action={action}
      ></NHSNotifyBackButton>
      <div className='nhsuk-grid-column-two-thirds'>
        <ReviewTemplate
          templateName={nhsAppTemplateName}
          sectionHeading={previewNHSAppFormComponent.sectionHeading}
          details={previewNHSAppFormComponent.details}
          form={{
            ...previewNHSAppFormComponent.form,
            state,
            action,
            formId: 'review-nhs-app-template',
            radiosId: 'reviewNHSAppTemplateAction',
            legend: {
              isPgeHeading: false,
              size: 'm',
            },
          }}
          PreviewComponent={<PreviewTemplate.NHSApp message={html} />}
        />
      </div>
    </div>
  );
}
