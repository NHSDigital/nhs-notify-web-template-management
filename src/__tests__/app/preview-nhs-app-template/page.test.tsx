import PreviewNhsAppTemplatePage from '@app/preview-nhs-app-template/[templateId]/page';
import { ReviewNHSAppTemplate } from '@forms/ReviewNHSAppTemplate/ReviewNHSAppTemplate';
import { NHSAppTemplate, TemplateType } from '@utils/types';
import { redirect } from 'next/navigation';
import { getTemplate } from '@utils/form-actions';

jest.mock('@utils/form-actions');
jest.mock('next/navigation');
jest.mock('@forms/ReviewNHSAppTemplate/ReviewNHSAppTemplate');
jest.mock('@utils/logger');

const redirectMock = jest.mocked(redirect);
const getTemplateMock = jest.mocked(getTemplate);

describe('PreviewNhsAppTemplatePage', () => {
  beforeEach(jest.resetAllMocks);

  it('should load page', async () => {
    const state: NHSAppTemplate = {
      id: 'template-id',
      version: 1,
      templateType: TemplateType.NHS_APP,
      NHS_APP: {
        name: 'template-name',
        message: 'template-message',
      },
    };

    getTemplateMock.mockResolvedValueOnce(state);

    const page = await PreviewNhsAppTemplatePage({
      params: {
        templateId: 'template-id',
      },
    });

    expect(page).toEqual(<ReviewNHSAppTemplate initialState={state} />);
  });

  it('should redirect to invalid-template when no template is found', async () => {
    await PreviewNhsAppTemplatePage({
      params: {
        templateId: 'template-id',
      },
    });

    expect(redirectMock).toHaveBeenCalledWith('/invalid-template', 'replace');
  });

  test.each([
    {
      templateType: TemplateType.LETTER,
      NHS_APP: {
        name: 'template-name',
        message: 'template-message',
      },
    },
    {
      templateType: TemplateType.EMAIL,
      NHS_APP: {
        name: 'template-name',
        message: 'template-message',
      },
    },
    {
      templateType: TemplateType.SMS,
      NHS_APP: {
        name: 'template-name',
        message: 'template-message',
      },
    },
    {
      templateType: TemplateType.NHS_APP,
      NHS_APP: {
        name: 'template-name',
        message: undefined as unknown as string,
      },
    },
    {
      templateType: TemplateType.NHS_APP,
      NHS_APP: {
        name: undefined as unknown as string,
        message: 'template-message',
      },
    },
    {
      templateType: TemplateType.NHS_APP,
      NHS_APP: {
        name: null as unknown as string,
        message: null as unknown as string,
      },
    },
  ])(
    'should redirect to invalid-template when template is $templateType and name is $nhsAppTemplateName and message is $nhsAppTemplateMessage',
    async (value) => {
      getTemplateMock.mockResolvedValueOnce({
        id: 'template-id',
        version: 1,
        ...value,
      });

      await PreviewNhsAppTemplatePage({
        params: {
          templateId: 'template-id',
        },
      });

      expect(redirectMock).toHaveBeenCalledWith('/invalid-template', 'replace');
    }
  );
});
