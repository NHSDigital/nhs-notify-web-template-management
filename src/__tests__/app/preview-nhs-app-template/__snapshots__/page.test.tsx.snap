// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PreviewNhsAppTemplatePage should load page 1`] = `
<ReviewNHSAppTemplate
  initialState={
    {
      "id": "session-id",
      "nhsAppTemplateMessage": "template-message",
      "nhsAppTemplateName": "template-name",
      "templateType": "NHS_APP",
    }
  }
/>
`;
